geopandas/geoseries.py

```diff
@@ -208,9 +208,16 @@ def __init__(self, data=None, index=None, crs: Optional[Any] = None, **kwargs):
                         "Non geometry data passed to GeoSeries constructor, "
                         f"received data of dtype '{s.dtype}'"
                     )
-            # try to convert to GeometryArray, if fails return plain Series
+            # extract object-dtype numpy array from pandas Series; with CoW this
+            # gives a read-only array, so we try to set the flag back to writeable
+            data = s.to_numpy()
             try:
-                data = from_shapely(s.values, crs)
+                data.flags.writeable = True
+            except ValueError:
+                pass
```

Comment:
I'm not sure when this might fail, but would we not need to make an explicit copy if setting the writeable flag fails - otherwise a version of martin's original issue happens?

Reply:
I should have added a test for the original issue, so that's working now.

In general, setting the `writeable` flag can fail when numpy doesn't "own" the memory. For example when a numpy array is created zero-copy from an array from another library like pytorch or pyarrow. At that point numpy refuses to change the read-only flag if that was set by the other library. Example with pyarrow, which assumes its arrays to be immutable and thus sets the writeable flag to false:

```
In [8]: arr = pa.array([1, 2, 3])

In [9]: np_arr = arr.to_numpy()

In [10]: np_arr.flags.writeable
Out[10]: False

In [11]: np_arr.flags.writeable = True
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[11], line 1
----> 1 np_arr.flags.writeable = True

ValueError: cannot set WRITEABLE flag to True of this array
```

Now, in practice we always have object dtype arrays, and I am not sure this will typically happen for this dtype (that's quite specific for numpy)

But so, when this happens, setting a value into the series fails with pandas as well. Continuing the example above:

```
In [15]: s = pd.Series(np_arr, copy=False)

In [16]: s.iloc[0] = 10
...
File ~/scipy/pandas/pandas/core/internals/blocks.py:1392, in Block.setitem(self, indexer, value)
   1390     casted = casted[0, ...]
   1391 try:
-> 1392     values[indexer] = casted
   1393 except (TypeError, ValueError) as err:
   1394     if is_list_like(casted):

ValueError: assignment destination is read-only
```

I _think_ it might be better for pandas to make this work (and perform the copy at the moment the setitem fails), but that's probably something for pandas to decide, and in geopandas we can just follow that.

Reply:
I may still be missing something, but it doesn't make sense to me that we catch the ValueError
```
            try:
                data.flags.writeable = True
            except ValueError:
                pass

```
and not copy data, since the subsequent `from_shapely` call will fail as well in that situation (or maybe I've got this wrong). 

In any case I'm happy to merge as I think this is only a theoretical issue - we'd need to recieve a numpy array of object dtype containing shapely objects which is also not possible to make writeable - and I'm not sure how to contrive that situation since arrow/ jax (and I think pytorch too) don't support python object arrays